File: 0239.SlidingWindowMaximum.go
func maxSlidingWindow(nums []int, k int) []int {
    n := len(nums)
    if n < k || k <= 0 {
        return nil
    }
    maxL, maxR := make([]int, n), make([]int, n)
    for i := 0; i < n; i++ {
        if i%k == 0 {
            maxL[i] = nums[i]
        } else {
            maxL[i] = max(maxL[i-1], nums[i])
        }
    }
    maxR[n-1] = nums[n-1]
    for i := n-2; i >= 0; i-- {
        if i%k == 0 {
            maxR[i] = nums[i]
        } else {
            maxR[i] = max(maxR[i+1], nums[i])
        }
    }
    ans := make([]int, n-k+1)
    for i := 0; i <= n-k; i++ {
        ans[i] = max(maxR[i], maxL[i+k-1])
    }
    return ans
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}

File: 0000-0999/0239.SlidingWindowMaximum.py
class Solution(object):
    def maxSlidingWindow(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        if len(nums) < k or k == 0:
            return []
        queue = []
        ans = []
        for i in xrange(len(nums)):
            while queue and nums[queue[-1]] < nums[i]:
                queue.pop()
            queue.append(i)
            if i < k-1:
                continue
            if queue and queue[0] <= i-k:
                queue.pop(0)
            ans.append(nums[queue[0]])
        return ans
